xarray/backends/api.py

```diff
@@ -876,16 +876,23 @@ def to_zarr(dataset, store=None, mode='w-', synchronizer=None, group=None,
     _validate_dataset_names(dataset)
     _validate_attrs(dataset)
 
-    store = backends.ZarrStore.open_group(store=store, mode=mode,
+    zstore = backends.ZarrStore.open_group(store=store, mode=mode,
                                           synchronizer=synchronizer,
                                           group=group)
 
     writer = ArrayWriter()
     # TODO: figure out how to properly handle unlimited_dims
-    dump_to_store(dataset, store, writer, encoding=encoding)
+    dump_to_store(dataset, zstore, writer, encoding=encoding)
     writes = writer.sync(compute=compute)
 
     if not compute:
         import dask
-        return dask.delayed(_finalize_store)(writes, store)
-    return store
+        return dask.delayed(_finalize_store)(writes, zstore)
+        # TODO: handle consolidation for delayed case
```

Comment:
This is the one major issue I don't know how to deal with. `consolidate_metadata` needs to be called on the target PutableMapping *after* the store is done writing. I don't know how to do this in a delayed context.

Reply:
what does the call signature look like for the consolidation step? We can probably make that a delayed thing and have it depend on the output from `_finalize_store`.

So something like:

```Python
@delayed
def consolidate_zarr(store, finalized_store):
    # replace store with whatever you need to pass to consolidate metadata
    return zarr.consolidate_metadata(store)

...
if not compute:
    import dask
    finalized_store = dask.delayed( _finalize_store)(writes, zstore)
    return consolidate_zarr(store, finalized_store)
```

`finalized_store` is bad name here but this is just a dummy argument to add the task dependency.

Reply:
the call signature is exactly as you said:
```
        import zarr
        zarr.consolidate_metadata(store)
```

Reply:
If you look at `_finalize_store`, it's just a few lines:
```python
def _finalize_store(write, store):
    """ Finalize this store by explicitly syncing and closing"""
    del write  # ensure writing is done first
    store.close()
```

So logically, it could make sense to do this cleanup in a `ZarrStore.close()` method, e.g.,
```
def close(self):
    if self._consolidate:
        zarr.consolidate_metadata(self.ds)
```

Reply:
Consolidate needs to be called after the metadata files are written, which must happen before the first data write, no? The remaining data writes can happen later, and data-sets opened via consolidated remain writeabel as far as data is concerned.

Reply:
Ah so you're saying we could just call consolidate before doing any writes? That would certainly be simpler.

Reply:
I *think* so, creating the array hierarchy should make the metadata without the need for any data files.